src/python/WMCore/Storage/RucioFileCatalog.py

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python
+"""
+_RucioFileCatalog_
+
+Object to contain LFN to PFN mappings from a Rucio File Catalog
+and provide functionality to match LFNs against them
+
+Usage:
+
+given a storage description file, storage.json, execute readRFC. This will return
+a RucioFileCatalog instance that can be used to match LFNs to PFNs.
+"""
+
+from builtins import next, str, range
+from future.utils import viewitems
+
+from future import standard_library
+standard_library.install_aliases()
+
+import os
+import re
+
+import json
+
+from urllib.parse import urlsplit
+from xml.dom.minidom import Document
+
+from WMCore.Algorithms.ParseXMLFile import xmlFileToNode
+
+class RucioFileCatalog(dict):
+    """
+    _RucioFileCatalog_
+
+    Object that can map LFNs to PFNs based on contents of a Rucio
+    File Catalog
+    """
+
+    def __init__(self):
+        dict.__init__(self)
+        self['lfn-to-pfn'] = []
+        self['pfn-to-lfn'] = []
+        self.preferredProtocol = None  # attribute for preferred protocol
+
+    def addMapping(self, protocol, match, result,
+                   chain=None, mapping_type='lfn-to-pfn'):
+        """
+        _addMapping_
+
+        Add an lfn to pfn mapping to this instance
+
+        """
+        entry = {}
+        entry.setdefault("protocol", protocol)
+        entry.setdefault("path-match-expr", re.compile(match))
+        entry.setdefault("path-match", match)
+        entry.setdefault("result", result)
+        entry.setdefault("chain", chain)
+        self[mapping_type].append(entry)
+
+    def _doMatch(self, protocol, path, style, caller):
+        """
+        Generalised way of building up the mappings.         
+        :param protocol: the name of a protocol, for example XRootD
+        :path: a LFN path, for example /store/abc/xyz.root
+        :style: type of conversion. lfn-to-pfn is to convert LFN to PFN and pfn-to-pfn is for PFN to LFN
+        :caller is the method from there this method was called. It's used for resolving chained rules
+        """
+        for mapping in self[style]:
+            if mapping['protocol'] != protocol:
+                continue
+            if mapping['path-match-expr'].match(path) or mapping["chain"] != None:
+                if mapping["chain"] != None:
+                    oldpath = path
+                    path = caller(mapping["chain"], path)
+                    if not path:
+                        continue
+                splitList = []
+                if len(mapping['path-match-expr'].split(path, 1)) > 1:
+                    for split in range(len(mapping['path-match-expr'].split(path, 1))):
+                        s = mapping['path-match-expr'].split(path, 1)[split]
+                        if s:
+                            splitList.append(s)
+                else:
+                    path = oldpath
+                    continue
+                result = mapping['result']
+                for split in range(len(splitList)):
```

Comment:
I am not very sure what these codes do. They are transferred directly from TrivialFileCatalog (master) as I mentioned above so I supposed it does the jobs. In general, the _doMatch method is transferred directly from TrivailFileCatalog (master) and I just add docstring to clarify what I understand

Reply:
Thanks Duong. Please don't dive in in any of these code that you have simply moved within WMCore.
If there is any module that you barely modified and reused most of it, you might want to add it to the PR description, such that we can just have a basic look at the implementation instead of deep diving into the how/why/what for of the implementation.

Reply:
Yes, I will add them to the PR description